lib/tokens.ts

```diff
@@ -0,0 +1,148 @@
+import { PublicKey } from "@solana/web3.js";
+import { TokenProps } from "./types";
+import {
+  JsonMetadata,
+  MPL_TOKEN_METADATA_PROGRAM_ID,
+  deserializeMetadata,
+} from "@metaplex-foundation/mpl-token-metadata";
+import { RpcAccount } from "@metaplex-foundation/umi";
+/**
+ * Starts with the jup.ag strict list to find token. jup.ag maintains a list of quality tokens
+ * if that fails, use metaplex with RPC call and fetch metadata json
+ * if THAT fails, take MINT address and truncate by 5 characters and that's the symbol (no image)
+ */
+export async function enrichTokenMetadata(
+  tokenAddress: PublicKey
+): Promise<TokenProps> {
+  // first check strict list
+  const tokenOnJup = await getTokenFromJupStrictList(tokenAddress);
+  if (tokenOnJup) {
+    return {
+      symbol: tokenOnJup.symbol,
+      publicKey: tokenAddress.toString(),
+      url: tokenOnJup.logoURI,
+      decimals: tokenOnJup.decimals,
+    };
+  }
+
+  // try metaplex
+  const jsonMetadata = await getMetaplexMetadataForToken(tokenAddress);
+  if (jsonMetadata && jsonMetadata.symbol) {
+    return {
+      symbol: jsonMetadata.symbol,
+      publicKey: tokenAddress.toString(),
+      url: jsonMetadata.image,
+      decimals: jsonMetadata.seller_fee_basis_points,
+    };
+  }
+
+  // finally just return truncated address for symbol
+  return {
+    symbol: tokenAddress.toString().slice(0, 5).toUpperCase(),
+    publicKey: tokenAddress.toString(),
```

Comment:
We still want to fetch decimals if we CAN and I THINK we can load in the spl-token program and have it do the gPA request for us too...

So in theory we want spl-token program built through Anchor (using like we have Program<>) and then one for token-2022.

Reply:
Hmm okay I'll look into this.

Reply:
Added the call to the Token 2022 program!